frame/offences/src/lib.rs

```diff
@@ -216,32 +302,51 @@ struct TriageOutcome<T: Config> {
 #[must_use = "The changes are not saved without called `save`"]
 struct ReportIndexStorage<T: Config, O: Offence<T::IdentificationTuple>> {
 	opaque_time_slot: OpaqueTimeSlot,
+	session_index: SessionIndex,
 	concurrent_reports: Vec<ReportIdOf<T>>,
+	session_reports: Vec<(Kind, OpaqueTimeSlot, ReportIdOf<T>)>,
 	_phantom: PhantomData<O>,
 }
 
 impl<T: Config, O: Offence<T::IdentificationTuple>> ReportIndexStorage<T, O> {
 	/// Preload indexes from the storage for the specific `time_slot` and the kind of the offence.
-	fn load(time_slot: &O::TimeSlot) -> Self {
+	fn load(time_slot: &O::TimeSlot, session_index: SessionIndex) -> Self {
 		let opaque_time_slot = time_slot.encode();
 
+		let session_reports = get_session_reports::<T>(session_index);
 		let concurrent_reports = <ConcurrentReportsIndex<T>>::get(&O::ID, &opaque_time_slot);
 
-		Self { opaque_time_slot, concurrent_reports, _phantom: Default::default() }
+		Self {
+			opaque_time_slot,
+			session_index,
+			concurrent_reports,
+			session_reports,
+			_phantom: Default::default(),
+		}
 	}
 
 	/// Insert a new report to the index.
-	fn insert(&mut self, report_id: ReportIdOf<T>) {
+	fn insert(&mut self, report_id: ReportIdOf<T>, time_slot: &O::TimeSlot) {
+		let opaque_time_slot = time_slot.encode();
+		self.session_reports.push((O::ID, opaque_time_slot, report_id));
+
 		// Update the list of concurrent reports.
 		self.concurrent_reports.push(report_id);
 	}
 
 	/// Dump the indexes to the storage.
 	fn save(self) {
+		SessionReports::<T>::set(self.session_index, self.session_reports.encode());
 		<ConcurrentReportsIndex<T>>::insert(
 			&O::ID,
 			&self.opaque_time_slot,
 			&self.concurrent_reports,
 		);
 	}
 }
+
+fn get_session_reports<T: Config>(session_index: SessionIndex) -> Vec<SessionReportOf<T>> {
+	let session_reports = SessionReports::<T>::get(session_index);
+	Vec::<SessionReportOf<T>>::decode(&mut &session_reports[..])
+		.unwrap_or_default()
```

Comment:
This  `unwrap_or_default` also seems to ignore the fact that either: There is nothing in storage and it should therefore return `vec![]` *or* there is corrupted storage.  
`defensive_unwrap_or_default` should be better. Or you return an error and bubble it up :man_shrugging:

Reply:
It's actually fine to have an empty vec here as a session might not have any reports.